### PR TITLE
feat(avalanche): integrate avalanche network 

### LINKED ISSUE #4864: [Bounty] Integrate Avalanche into Mask
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  [x] 4 weeks

- Bounty Size
  [x] 5000 USD

### Part 1: Problem Description

**Background** Avalanche is a flexible smart contract platform for creating custom blockchain networks and decentralized applications (especially those focused on DeFi). Avalanche facilitates the creation of custom blockchains (including private blockchains) and decentralized applications (dApps). Its native,fixed-supply AVAX coin is the platform’s unit of account, used for paying transaction fees and for staking in Avalanche’s consensus process.

**Solution** As the first step of our cooperation, we, Mask Network, would like to integrate Avalanche as one of our default supported blockchain system in our internal Mask Wallet, so that further on-chain interactions would happen directly within Mask Network without leaving the platform. We believe this would further help us attract more users from different projects, aiming at different goals to maximize users experiences in the broad blockchain world.
If you are interested in this bounty, please make sure that you contact us first to see how our structure looks like and proceed to the development afterwards.

### Part 2: Resources

- https://www.avax.network/
- https://github.com/DimensionDev/Maskbook/tree/master/packages/maskbook/src/plugins/Wallet

### Part 3: Contact and Support

@yisiliu for ideas @Jack-Works for implementations Mask Network discord server: https://discord.gg/4Xccj7a75Q Avalanche discord server: https://discord.com/invite/RwXY7P6

### PR DESCRIPTION
**Pull request by developerfred**
## Description
Integrate Avalanche Network 

Closes #4864, #4874

## Type of change

- [x]  add sushiswap on Avalanche 
- [x] add traderJoe DEX on Avalanche 
- [x] add Pangolin Dex on Avalanche 
- [x]  add avalanche network

#### NFT Red Packet 
- 02fe77d5c  [0x3E99d32B8c601A6628d301673bC28B17F2aDB326](https://snowtrace.io/tx/0xe5eedfd20fac7b13d109902ae2dbdfc9a0ac0f2b5300a8aaca9cd2a6e855a1e0) 

#### RedPacket 
- [0x525Da0e87199940EDea5BfB6D9c057333D915F57](https://snowtrace.io/tx/0x725bdd0c60d398f0e5d68229aed8db18acb629c0d887c0d078c06354cb36f993)
#### ITO 
- [0x452F3023f0E163056bEFFFcA719B12e0dD7Af451](https://snowtrace.io/address/0x452F3023f0E163056bEFFFcA719B12e0dD7Af451)
#### MaskBox 
- [0x0477524B070F73BE7f84f3326d104899e1beBb12](https://snowtrace.io/address/0x0477524B070F73BE7f84f3326d104899e1beBb12)

- [x] https://github.com/DimensionDev/NFT-RedPacket-Subgraph/pull/3
- [x] https://github.com/DimensionDev/SmartContractProjectConfig/pull/7

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

**Review by septs — CHANGES_REQUESTED** [review]
plz resolve build failed, thanks

**Comment by developerfred** [comment]
> plz resolve build failed, thanks

I'm fixing this.

File: packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx
Comment (septs) [inline packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx]:
```suggestion
            onChange: (value: string) => Services.Settings.setAvalancheNetworkTradeProvider(Number.parseInt(value, 10)),
```

Comment (developerfred) [reply packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx]: thx, fixed.

**Comment by developerfred** [comment]
@septs  I'm very upset with this error **"merge_conflict_check"** 

I believe the solution is this: https://dev.to/callmekatootie/debug-resource-not-accessible-by-integration-error-when-working-with-githubs-graphql-endpoint-5bim

**Comment by developerfred** [comment]
@septs what do you think about this https://github.com/DimensionDev/Maskbook/pull/5289#issuecomment-1002805314 ?

**Comment by developerfred** [comment]
@septs done!

**Comment by septs** [comment]
> @septs I'm very upset with this error **"merge_conflict_check"**
> 
> I believe the solution is this: https://dev.to/callmekatootie/debug-resource-not-accessible-by-integration-error-when-working-with-githubs-graphql-endpoint-5bim

I'm design new solution replace the github action bot (merge_conflict_check)

https://github.com/DimensionDev/detect-pull-request-conflict

**Comment by developerfred** [comment]
> > @septs I'm very upset with this error **"merge_conflict_check"**
> > I believe the solution is this: [dev.to/callmekatootie/debug-resource-not-accessible-by-integration-error-when-working-with-githubs-graphql-endpoint-5bim](https://dev.to/callmekatootie/debug-resource-not-accessible-by-integration-error-when-working-with-githubs-graphql-endpoint-5bim)
> 
> I'm design new solution replace the github action bot (merge_conflict_check)
> 
> [DimensionDev/detect-pull-request-conflict](https://github.com/DimensionDev/detect-pull-request-conflict)

Thanks @septs

**Comment by developerfred** [comment]
@septs  @guanbinrui

**Comment by developerfred** [comment]
Screenshot 2022-01-05 at 04 46 28
Screenshot 2022-01-05 at 04 46 19

File: packages/mask/src/components/shared/ApplicationBoard.tsx
Comment (septs) [inline packages/mask/src/components/shared/ApplicationBoard.tsx]: https://github.com/DimensionDev/Maskbook/pull/5398

Comment (developerfred) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: @septs I'll wait for the merge then update

Comment (developerfred) [reply packages/mask/src/components/shared/ApplicationBoard.tsx]: done here 863cc8a

**Comment by wenluomask** [comment]
Thanks for your effort, there are some issues I found in this latest pr.
- Dashboard：
1. Asset page didn't show the balance of Avalanche network, the data should be fetch from DeBank;
Screen Shot 2022-01-10 at 6 21 56 PM

2. The background of current network should be red, currently is white;
Screen Shot 2022-01-10 at 6 27 44 PM

3. Click swap button in Token Asset page, currently it will open Arbitrum's swap page. It should be Avalanche's swap page and  choose the default input and output token;
Screen Shot 2022-01-11 at 11 24 08 PM

-DEX:
4. The swap pair works in Pangolin, but it doesn't work in Swap. It was showing Insufficient liquiditity;
Screen Shot 2022-01-11 at 11 29 49 PM
Screen Shot 2022-01-11 at 11 29 59 PM

5. Estimate US dollar value of token is missing in Swap or Popup;
Screen Shot 2022-01-11 at 11 35 01 PM

-SNS:
6. Unable to create lucky drop or ITO in Avalanche network;

**Comment by developerfred** [comment]
@wenluomask I will work on this review

**Comment by developerfred** [comment]
5. ✅  Estimate US dollar value of token is missing in Swap or Popup;
bug on coingecko api https://www.reddit.com/r/coingecko/comments/kxpfju/coingecko_api_broken_for_avax/
fix: [binance-peg-avalanche](https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&symbols=avax)

Screenshot 2022-01-11 at 23 59 34

File: packages/mask/src/resources/PangolinIcon.tsx
Comment (septs) [inline packages/mask/src/resources/PangolinIcon.tsx:4]: plz use https://aka.pw/svgomg reduce this svg image

Comment (developerfred) [reply packages/mask/src/resources/PangolinIcon.tsx:4]: 2.73k → 1.35k

**Comment by developerfred** [comment]
fixed tranding view swap for avalanche ($avax)
Screenshot 2022-01-12 at 00 42 02

**Comment by developerfred** [comment]
@wenluomask  about 6. Unable to create lucky drop or ITO in Avalanche network;
I will need gas to create the contracts, I requested the team through the telegram group.
cc: @LindaMask @yisiliu

**Comment by developerfred** [comment]
- 1  (deBank api)
- 2 ✅ 
- 3 ✅ 
- 4 ✅ 
- 5 ✅ 
- 6 ✅

**Comment by wenluomask** [comment]
> 

Hi @developerfred , our team will handle this. So don't worry about #6.

**Comment by developerfred** [comment]
> > 
> 
> Hi @developerfred , our team will handle this. So don't worry about #6.

6 ✅

**Comment by developerfred** [comment]
-4 ✅  @wenluomask
Screenshot 2022-01-14 at 15 29 33
Screenshot 2022-01-14 at 15 29 55

**Comment by developerfred** [comment]
-> https://youtu.be/W-Ar8TcLbBI

**Comment by wenluomask** [comment]
First, I hope you can merge latest develop into this pr to update. And here is some feedback:

- [x] 1. Debank api should be fine, but I still cannot see any balance in Asset page;
Screen Shot 2022-01-14 at 6 06 02 PM

2&3 are good to go;

- [x] 4. Why do we need to wrap AVAX first? In Pangolin I can swap AVAX to other tokens directly;

- [x] 5. Estimate US dollar value of token is still missing in Popup wallet;

6. Issue about Lucky Drop and ITO:(@guanbinrui @zhouhanseng )

- [x] -If claim or create a Lucky Drop or ITO, the max priority fee and max fee sent to wallet were not the base fee;

Screen Shot 2022-01-17 at 1 17 23 PM
Screen Shot 2022-01-18 at 6 07 51 PM
Here is the tx:
https://snowtrace.io/tx/0x10565f74a3a395ec4ccc0d54a4747e0282be87c84efe5835c8ac5acbe50671d2

- [x] -Claiming status doesn't work in this pr;

- [x] -ITO is not supported yet;

**Comment by developerfred** [comment]
> 4\. Why do we need to wrap AVAX first? In Pangolin I can swap AVAX to other tokens directly;

@wenluomask Pangolin has no default liquidity for Avax only for the WAVAX envelope

**Comment by developerfred** [comment]
ITO ✅
4d259f3 https://github.com/DimensionDev/InitialTwitterOffering/tree/add_more_networks

**Comment by wenluomask** [comment]
7. 4 is fixed by commit fix: fix swap, but we still need the estimate gas fee in DEX card.
Screen Shot 2022-01-18 at 5 47 53 PM

8. AVAX is not set in Claim page yet;( Currently there is the tab, but I cannot see the token claimable.)
Screen Shot 2022-01-18 at 6 10 50 PM

9. Token icons for AVAX's token will be missing in token list.
Screen Shot 2022-01-18 at 6 06 50 PM
Screen Shot 2022-01-18 at 6 02 04 PM

**Comment by developerfred** [comment]
AVAX is not set in Claim page yet; 
@zhouhanseng  @UncleBill  Has the contract already been implemented?

**Comment by UncleBill** [comment]
> AVAX is not set in Claim page yet; @zhouhanseng @UncleBill Has the contract already been implemented?

Here the contract is. https://github.com/DimensionDev/InitialTwitterOffering/tree/add_more_networks

**Comment by UncleBill** [comment]
@wenluomask 

Icons fixed
image

tab fixed
image

**Comment by developerfred** [comment]
@wenluomask is there a way for me to test these ITO contracts? or only who is the owner of the contract can create an ITO?

Everyone can create an ITO, you can select an inexpensive token to try it.

**Comment by wenluomask** [comment]
> @wenluomask
> 
> Icons fixed image
> 
> tab fixed image

Currently there is the tab, but I cannot see the token claimable.

**Comment by developerfred** [comment]
> > @wenluomask
> > Icons fixed image
> > tab fixed image
> 
> Currently there is the tab, but I cannot see the token claimable.

I asked @UncleBill  for the implementation and he informed me that the Airdrop contracts are depracated, @wenluomask what is the next step to merge?

**Comment by developerfred** [comment]
> > > @wenluomask
> > > Icons fixed image
> > > tab fixed image
> > 
> > 
> > Currently there is the tab, but I cannot see the token claimable.
> 
> I asked @UncleBill for the implementation and he informed me that the Airdrop contracts are depracated, @wenluomask what is the next step to merge?

@LindaMask look here

**Review by Jack-Works — DISMISSED** [review]

**Comment by UncleBill** [comment]
Submitted a PR to https://github.com/trustwallet/assets/pull/17549 to fix the missing token icon.

**Comment by UncleBill** [comment]
Use logo from our forked repo

image

**Comment by developerfred** [comment]
> Use logo from our forked repo
> 
> image

done here: 44d98ca

**Comment by UncleBill** [comment]
@developerfred use this repo once this PR gets merged.

**Comment by developerfred** [comment]
@UncleBill @wenluomask 

**about 7**

The Merge can be carried forward, from my investigation the problem is in an overload of the avalanche of requests, so I didn't show the gas price of the other DEXs the only dex that took this into account when implementing their contracts was sushiswap.

We can create a hotfix in the future thinking about a customizable gas limit, because this will change the project and influence other chains.

TL;DR the same thing that happens in the avalanche happens in the Fantom network, however in the fantom not even the sushiswap can calculate the gas

Avalanche has stable gas cost -> https://snowtrace.io/gastracker

Screenshot 2022-01-21 at 15 35 53
Screenshot 2022-01-21 at 13 51 31

**Comment by developerfred** [comment]
the crazy thing about it all is that gas appears to unwrap it.

**Comment by developerfred** [comment]
the crazy thing about it all is that gas appears to unwrap it.
Screenshot 2022-01-21 at 16 13 57
Screenshot 2022-01-21 at 16 15 25

**Review by UncleBill — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Comment by developerfred** [comment]
@guanbinrui @DimensionDev/qa

**Review by Jack-Works — DISMISSED** [review]

**Review by UncleBill — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Comment by wenluomask** [comment]
Checked.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**
